x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java

```diff
@@ -280,22 +283,41 @@ static void validateExpiredTime(final Date exp, final Date now, final long allow
 
+        // If JWT has optional kid header, and realm JWKs have optional kid attribute, any mismatches JWT.kid vs JWK.kid can be ignored.
+        // Keep any JWKs if JWK optional kid attribute is missing. Keep all JWKs if JWT optional kid header is missing.
         final List<JWK> jwksKid = jwks.stream().filter(j -> ((id == null) || (j.getKeyID() == null) || (id.equals(j.getKeyID())))).toList();
-        LOGGER.trace("JWKs [" + jwksKid.size() + "] after KID [" + id + "||null] filter.");
+        LOGGER.debug("JWKs [" + jwksKid.size() + "] after KID [" + id + "||null] filter.");
 
+        // JWT has mandatory alg header. If realm JWKs have optional alg attribute, any mismatches JWT.alg vs JWK.alg can be ignored.
+        // Keep any JWKs if JWK optional alg attribute is missing.
         final List<JWK> jwksAlg = jwksKid.stream().filter(j -> (j.getAlgorithm() == null) || (alg.equals(j.getAlgorithm()))).toList();
-        LOGGER.trace("JWKs [" + jwksAlg.size() + " after Algorithm [" + alg.getName() + "||null] filter.");
+        LOGGER.debug("JWKs [" + jwksAlg.size() + " after Algorithm [" + alg.getName() + "||null] filter.");
 
+        // PKC Example: Realm has five PKC JWKs RSA-2048, RSA-3072, EC-P256, EC-P384, and EC-P512. JWT alg allows ignoring some.
+        // - If JWT alg is RS256, only RSA-2048 and RSA-3072 are valid for a JWT RS256 signature. Ignore three EC JWKs.
+        // - If JWT alg is ES512, only EC-P512 is valid for a JWT ES512 signature. Ignore four JWKs (two RSA, two EC).
+        // - If JWT alg is ES384, only EC-P384 is valid for a JWT ES384 signature. Ignore four JWKs (two RSA, two EC).
+        // - If JWT alg is ES256, only EC-P256 is valid for a JWT ES256 signature. Ignore four JWKs (two RSA, two EC).
+        //
+        // HMAC Example: Realm has six HMAC JWKs of bit lengths 256, 320, 384, 400, 512, and 1000. JWT alg allows ignoring some.
+        // - If JWT alg is HS256, all are valid for a JWT HS256 signature. Don't ignore any HMAC JWKs.
+        // - If JWT alg is HS384, only 384, 400, 512, and 1000 are valid for a JWT HS384 signature. Ignore two HMAC JWKs.
+        // - If JWT alg is HS512, only 512 and 1000 are valid for a JWT HS512 signature. Ignore four HMAC JWKs.
         final List<JWK> jwksStrength = jwksAlg.stream().filter(j -> JwkValidateUtil.isMatch(j, alg.getName())).toList();
-        LOGGER.trace("JWKs [" + jwksStrength.size() + "] after Algorithm [" + alg + "] match filter.");
+        LOGGER.debug("JWKs [" + jwksStrength.size() + "] after Algorithm [" + alg + "] match filter.");
 
-        for (final JWK jwk : jwksStrength) {
-            if (jwt.verify(JwtValidateUtil.createJwsVerifier(jwk))) {
-                return; // VERIFY SUCCEEDED
+        if (jwksStrength.isEmpty()) {
+            LOGGER.warn("JWKs [" + jwks.size() + "] usable [" + jwksStrength.size() + "]. See debug logs.");
+        } else {
```

Comment:
If the list is empty, does that mean authentication will sure fail?
Also, I'd drop "See debug logs" which assumes debug logs are readily available but that is unlikely true.

Reply:
Any of the filters could reduce the JWK list to zero. The strength filter is the last filter. If none left at that point, there is no reason to go through the for-loop.

Example:
- Realm has HMAC 256-bit key.
- Current request is a JWT signed with HS512 algorithm.
- Trying the HMAC 256-bit key is pointless, it is too short for HS512 algorithm.
- In other words, only HMAC keys of 512-bit or larger need to be tried for JWTs with HS512 algorithm. If all HMAC keys are smaller, the validate for-loop is empty, and we should log a warning to tell customers that all HS512 JWTs have no chance of ever being authenticated by the realm.

Reply:
Changed from warn to trace. Resolved.